File: Clausage/Services/UsageService.swift
import Foundation
import SwiftData

struct UsageData: Equatable {
    var fiveHourPercent: Double?
    var fiveHourResetsAt: Date?
    var weeklyPercent: Double?
    var weeklyResetsAt: Date?
    var lastUpdated: Date?
    var error: String?
    /// True when showing cached data after a transient API failure
    var isStale: Bool = false

    static func == (lhs: UsageData, rhs: UsageData) -> Bool {
        lhs.fiveHourPercent == rhs.fiveHourPercent
        && lhs.weeklyPercent == rhs.weeklyPercent
        && lhs.lastUpdated == rhs.lastUpdated
        && lhs.error == rhs.error
    }
}

@Observable
final class UsageService {
    var usage = UsageData()
    var isLoading = false

    private var refreshTimer: Timer?
    private var consecutiveFailures = 0
    private var lastSuccessfulUsage: UsageData?
    private var modelContainer: ModelContainer?
    private var retryTask: Task<Void, Never>?

    private static let cachedUsageKey = "lastSuccessfulUsage"

    init() {
        // Restore last known usage from disk so the user never sees an empty state
        if let cached = Self.loadCachedUsage() {
            var restored = cached
            restored.isStale = true
            usage = restored
            lastSuccessfulUsage = cached
        }
        startRefreshTimer()
        Task { @MainActor in
            self.fetch()
        }
    }

    func setModelContainer(_ container: ModelContainer) {
        self.modelContainer = container
    }

    private func startRefreshTimer() {
        refreshTimer?.invalidate()
        let interval = AppSettings.shared.refreshInterval
        refreshTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            let s = self
            Task { @MainActor in s?.fetch() }
        }
    }

    func updateRefreshInterval() {
        startRefreshTimer()
    }

    @MainActor
    func fetch() {
        isLoading = true
        // Cancel any pending retry to avoid compounding fetches
        retryTask?.cancel()
        retryTask = nil

        Task.detached(priority: .utility) {
            let result = Self.fetchUsage()

            await MainActor.run { [weak self] in
                guard let self else { return }

                let isRateLimited = result.error == "Rate limited"

                if result.error != nil {
                    self.consecutiveFailures += 1

                    if let lastGood = self.lastSuccessfulUsage {
                        var stale = lastGood
                        stale.isStale = true
                        self.usage = stale
                    } else if isRateLimited {
                        self.usage = UsageData(error: "Waiting for API... (rate limited)")
                    } else {
                        self.usage = result
                    }

                    // On rate limit: don't retry — wait for next regular refresh.
                    // On other errors: retry with backoff.
                    if !isRateLimited {
                        let delay = min(15.0 * pow(2.0, Double(self.consecutiveFailures - 1)), 120.0)
                        self.retryTask = Task { [weak self] in
                            try? await Task.sleep(for: .seconds(delay))
                            guard !Task.isCancelled else { return }
                            await self?.fetch()
                        }
                    }
                } else {
                    self.consecutiveFailures = 0
                    self.lastSuccessfulUsage = result
                    self.usage = result
                    self.persistSnapshot(result)
                    Self.saveCachedUsage(result)
                }
                self.isLoading = false
            }
        }
    }

    private func persistSnapshot(_ data: UsageData) {
        guard let container = modelContainer,
              let fiveHour = data.fiveHourPercent,
              let weekly = data.weeklyPercent else { return }

        let snapshot = UsageSnapshot(
            timestamp: Date(),
            fiveHourPercent: fiveHour,
            weeklyPercent: weekly,
            fiveHourResetsAt: data.fiveHourResetsAt,
            weeklyResetsAt: data.weeklyResetsAt
        )

        Task.detached {
            let context = ModelContext(container)
            context.insert(snapshot)
            try? context.save()
        }
    }

    // MARK: - API

    private static func fetchUsage() -> UsageData {
        guard let token = KeychainService.getAccessToken() else {
            return UsageData(error: "No Claude Code credentials found. Open Claude Code and log in first.")
        }

        guard let url = URL(string: "https://api.anthropic.com/api/oauth/usage") else {
            return UsageData(error: "Invalid URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("oauth-2025-04-20", forHTTPHeaderField: "anthropic-beta")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        // The usage API is undocumented and rate-limits unknown user agents.
        // Using the Claude Code UA is necessary until Anthropic provides a public usage API.
        request.setValue("claude-code/2.1.77", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 10

        let semaphore = DispatchSemaphore(value: 0)
        var responseData: Data?
        var responseError: Error?
        var httpStatusCode: Int?

        URLSession.shared.dataTask(with: request) { data, response, error in
            responseData = data
            responseError = error
            httpStatusCode = (response as? HTTPURLResponse)?.statusCode
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let error = responseError {
            return UsageData(error: error.localizedDescription)
        }

        if httpStatusCode == 401 {
            guard let freshToken = KeychainService.refreshToken() else {
                return UsageData(error: "Authentication failed. Re-login to Claude Code.")
            }
            request.setValue("Bearer \(freshToken)", forHTTPHeaderField: "Authorization")

            let retrySemaphore = DispatchSemaphore(value: 0)
            var retryData: Data?
            var retryError: Error?
            URLSession.shared.dataTask(with: request) { data, _, error in
                retryData = data
                retryError = error
                retrySemaphore.signal()
            }.resume()
            retrySemaphore.wait()

            if let error = retryError {
                return UsageData(error: error.localizedDescription)
            }
            guard let data = retryData else {
                return UsageData(error: "No data received")
            }
            return parseResponse(data)
        }

        if httpStatusCode == 429 {
            // Token may have been rotated — try a fresh one before giving up
            if let freshToken = KeychainService.refreshToken() {
                request.setValue("Bearer \(freshToken)", forHTTPHeaderField: "Authorization")

                let retrySemaphore = DispatchSemaphore(value: 0)
                var retryData: Data?
                var retryError: Error?
                var retryStatus: Int?
                URLSession.shared.dataTask(with: request) { data, response, error in
                    retryData = data
                    retryError = error
                    retryStatus = (response as? HTTPURLResponse)?.statusCode
                    retrySemaphore.signal()
                }.resume()
                retrySemaphore.wait()

                if retryError == nil, retryStatus == 200, let data = retryData {
                    return parseResponse(data)
                }
            }
            return UsageData(error: "Rate limited")
        }

        if let code = httpStatusCode, code != 200 {
            return UsageData(error: "API error (\(code))")
        }

        guard let data = responseData else {
            return UsageData(error: "No data received")
        }

        return parseResponse(data)
    }

    private static func parseResponse(_ data: Data) -> UsageData {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return UsageData(error: "Failed to parse response")
        }

        if let error = json["error"] as? [String: Any],
           let message = error["message"] as? String {
            return UsageData(error: message)
        }

        var usage = UsageData(lastUpdated: Date())

        let isoFormatter = ISO8601DateFormatter()
        isoFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]

        if let fiveHour = json["five_hour"] as? [String: Any] {
            usage.fiveHourPercent = parseDouble(fiveHour["utilization"])
            if let resetStr = fiveHour["resets_at"] as? String {
                usage.fiveHourResetsAt = isoFormatter.date(from: resetStr)
            }
        }

        if let sevenDay = json["seven_day"] as? [String: Any] {
            usage.weeklyPercent = parseDouble(sevenDay["utilization"])
            if let resetStr = sevenDay["resets_at"] as? String {
                usage.weeklyResetsAt = isoFormatter.date(from: resetStr)
            }
        }

        if usage.fiveHourPercent == nil && usage.weeklyPercent == nil {
            let rawStr = String(data: data, encoding: .utf8) ?? "unknown"
            usage.error = "Unexpected response: \(rawStr.prefix(200))"
        }

        return usage
    }

    private static func parseDouble(_ value: Any?) -> Double? {
        if let d = value as? Double { return d }
        if let i = value as? Int { return Double(i) }
        if let s = value as? String { return Double(s) }
        return nil
    }

    // MARK: - Usage cache (survives app restarts)

    private static func saveCachedUsage(_ data: UsageData) {
        let dict: [String: Any?] = [
            "fiveHourPercent": data.fiveHourPercent,
            "weeklyPercent": data.weeklyPercent,
            "fiveHourResetsAt": data.fiveHourResetsAt?.timeIntervalSince1970,
            "weeklyResetsAt": data.weeklyResetsAt?.timeIntervalSince1970,
            "lastUpdated": data.lastUpdated?.timeIntervalSince1970
        ]
        UserDefaults.standard.set(dict, forKey: cachedUsageKey)
    }

    private static func loadCachedUsage() -> UsageData? {
        guard let dict = UserDefaults.standard.dictionary(forKey: cachedUsageKey) else { return nil }

        var data = UsageData()
        data.fiveHourPercent = dict["fiveHourPercent"] as? Double
        data.weeklyPercent = dict["weeklyPercent"] as? Double

        if let ts = dict["fiveHourResetsAt"] as? Double {
            data.fiveHourResetsAt = Date(timeIntervalSince1970: ts)
        }
        if let ts = dict["weeklyResetsAt"] as? Double {
            data.weeklyResetsAt = Date(timeIntervalSince1970: ts)
        }
        if let ts = dict["lastUpdated"] as? Double {
            data.lastUpdated = Date(timeIntervalSince1970: ts)
        }

        // Only return if there's actual data
        guard data.fiveHourPercent != nil || data.weeklyPercent != nil else { return nil }
        return data
    }

    static func resetTimeString(_ date: Date?) -> String {
        guard let date = date else { return "\u{2014}" }
        let interval = date.timeIntervalSince(Date())
        if interval <= 0 { return "now" }

        let hours = Int(interval) / 3600
        let minutes = (Int(interval) % 3600) / 60

        if hours > 24 {
            let days = hours / 24
            return "in \(days)d \(hours % 24)h"
        } else if hours > 0 {
            return "in \(hours)h \(minutes)m"
        } else {
            return "in \(minutes)m"
        }
    }
}

File: windows/Clausage/Services/UsageService.cs
using System.Text.Json;
using Clausage.Models;

namespace Clausage.Services;

public class UsageService
{
    private const string ApiUrl = "https://api.anthropic.com/api/oauth/usage";

    public UsageData Usage { get; private set; } = new();
    public bool IsLoading { get; private set; }

    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(10) };
    private System.Threading.Timer? _refreshTimer;
    private System.Threading.Timer? _retryTimer;
    private int _consecutiveFailures;
    private UsageData? _lastSuccessful;
    private readonly object _lock = new();
    private readonly string _cachePath;

    public UsageService()
    {
        var dir = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".clausage");
        Directory.CreateDirectory(dir);
        _cachePath = Path.Combine(dir, "usage_cache.json");

        var cached = LoadCached();
        if (cached != null)
        {
            cached.IsStale = true;
            Usage = cached;
            _lastSuccessful = new UsageData
            {
                FiveHourPercent = cached.FiveHourPercent,
                FiveHourResetsAt = cached.FiveHourResetsAt,
                WeeklyPercent = cached.WeeklyPercent,
                WeeklyResetsAt = cached.WeeklyResetsAt,
                LastUpdated = cached.LastUpdated,
            };
        }

        StartRefreshTimer();
        Task.Run(() => Fetch());
    }

    private void StartRefreshTimer()
    {
        _refreshTimer?.Dispose();
        var interval = TimeSpan.FromSeconds(AppSettings.Shared.RefreshInterval);
        _refreshTimer = new System.Threading.Timer(_ => Fetch(), null, interval, interval);
    }

    public void UpdateRefreshInterval() => StartRefreshTimer();

    public void Fetch()
    {
        lock (_lock)
        {
            IsLoading = true;
            _retryTimer?.Dispose();
            _retryTimer = null;
        }

        var result = FetchUsage();
        var isRateLimited = result.Error == "Rate limited";

        lock (_lock)
        {
            if (result.Error != null)
            {
                _consecutiveFailures++;
                if (_lastSuccessful != null)
                {
                    Usage = new UsageData
                    {
                        FiveHourPercent = _lastSuccessful.FiveHourPercent,
                        FiveHourResetsAt = _lastSuccessful.FiveHourResetsAt,
                        WeeklyPercent = _lastSuccessful.WeeklyPercent,
                        WeeklyResetsAt = _lastSuccessful.WeeklyResetsAt,
                        LastUpdated = _lastSuccessful.LastUpdated,
                        IsStale = true,
                    };
                }
                else if (isRateLimited)
                {
                    Usage = new UsageData { Error = "Waiting for API... (rate limited)" };
                }
                else
                {
                    Usage = result;
                }

                if (!isRateLimited)
                {
                    var delay = Math.Min(15.0 * Math.Pow(2, _consecutiveFailures - 1), 120.0);
                    _retryTimer = new System.Threading.Timer(_ => Fetch(), null,
                        TimeSpan.FromSeconds(delay), Timeout.InfiniteTimeSpan);
                }
            }
            else
            {
                _consecutiveFailures = 0;
                _lastSuccessful = result;
                Usage = result;
                SaveCached(result);
            }

            IsLoading = false;
        }
    }

    private UsageData FetchUsage()
    {
        var token = CredentialService.GetAccessToken();
        if (token == null)
            return new UsageData { Error = "No Claude Code credentials found. Open Claude Code and log in first." };

        try
        {
            var request = CreateRequest(token);
            var resp = _http.Send(request);

            if ((int)resp.StatusCode == 401)
            {
                var fresh = CredentialService.RefreshToken();
                if (fresh == null)
                    return new UsageData { Error = "Authentication failed. Re-login to Claude Code." };
                request = CreateRequest(fresh);
                resp = _http.Send(request);
            }

            if ((int)resp.StatusCode == 429)
            {
                var fresh = CredentialService.RefreshToken();
                if (fresh != null)
                {
                    try
                    {
                        request = CreateRequest(fresh);
                        var retryResp = _http.Send(request);
                        if (retryResp.IsSuccessStatusCode)
                        {
                            var retryBody = retryResp.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                            return ParseResponse(retryBody);
                        }
                    }
                    catch { }
                }
                return new UsageData { Error = "Rate limited" };
            }

            if (!resp.IsSuccessStatusCode)
                return new UsageData { Error = $"API error ({(int)resp.StatusCode})" };

            var body = resp.Content.ReadAsStringAsync().GetAwaiter().GetResult();
            return ParseResponse(body);
        }
        catch (Exception ex)
        {
            return new UsageData { Error = ex.Message };
        }
    }

    private static HttpRequestMessage CreateRequest(string token)
    {
        var req = new HttpRequestMessage(HttpMethod.Get, ApiUrl);
        req.Headers.Add("Authorization", $"Bearer {token}");
        req.Headers.Add("anthropic-beta", "oauth-2025-04-20");
        req.Headers.Add("User-Agent", "claude-code/2.1.77");
        return req;
    }

    internal static UsageData ParseResponse(string body)
    {
        try
        {
            using var doc = JsonDocument.Parse(body);
            var root = doc.RootElement;

            if (root.TryGetProperty("error", out var err))
            {
                var msg = err.ValueKind == JsonValueKind.Object && err.TryGetProperty("message", out var m)
                    ? m.GetString() : err.ToString();
                return new UsageData { Error = msg ?? "Unknown error" };
            }

            var usage = new UsageData { LastUpdated = DateTime.UtcNow };

            if (root.TryGetProperty("five_hour", out var fh))
            {
                usage.FiveHourPercent = ParseNumber(fh, "utilization");
                if (fh.TryGetProperty("resets_at", out var ra))
                {
                    if (DateTime.TryParse(ra.GetString(), null,
                        System.Globalization.DateTimeStyles.RoundtripKind, out var dt))
                        usage.FiveHourResetsAt = dt.ToUniversalTime();
                }
            }

            if (root.TryGetProperty("seven_day", out var sd))
            {
                usage.WeeklyPercent = ParseNumber(sd, "utilization");
                if (sd.TryGetProperty("resets_at", out var ra))
                {
                    if (DateTime.TryParse(ra.GetString(), null,
                        System.Globalization.DateTimeStyles.RoundtripKind, out var dt))
                        usage.WeeklyResetsAt = dt.ToUniversalTime();
                }
            }

            if (usage.FiveHourPercent == null && usage.WeeklyPercent == null)
                usage.Error = $"Unexpected response: {body[..Math.Min(body.Length, 200)]}";

            return usage;
        }
        catch
        {
            return new UsageData { Error = "Failed to parse response" };
        }
    }

    private static double? ParseNumber(JsonElement parent, string property)
    {
        if (!parent.TryGetProperty(property, out var val)) return null;
        return val.ValueKind switch
        {
            JsonValueKind.Number => val.GetDouble(),
            JsonValueKind.String => double.TryParse(val.GetString(), System.Globalization.NumberStyles.Float, System.Globalization.CultureInfo.InvariantCulture, out var d) ? d : null,
            _ => null
        };
    }

    public static string ResetTimeString(DateTime? date)
    {
        if (date == null) return "\u2014";
        var interval = (date.Value.ToUniversalTime() - DateTime.UtcNow).TotalSeconds;
        if (interval <= 0) return "now";

        int hours = (int)interval / 3600;
        int minutes = ((int)interval % 3600) / 60;

        if (hours > 24)
        {
            int days = hours / 24;
            return $"in {days}d {hours % 24}h";
        }
        if (hours > 0) return $"in {hours}h {minutes}m";
        return $"in {minutes}m";
    }

    private void SaveCached(UsageData data)
    {
        try
        {
            var cache = new Dictionary<string, object?>
            {
                ["five_hour_percent"] = data.FiveHourPercent,
                ["weekly_percent"] = data.WeeklyPercent,
                ["five_hour_resets_at"] = data.FiveHourResetsAt?.ToString("O"),
                ["weekly_resets_at"] = data.WeeklyResetsAt?.ToString("O"),
                ["last_updated"] = data.LastUpdated?.ToString("O"),
            };
            File.WriteAllText(_cachePath, JsonSerializer.Serialize(cache));
        }
        catch { }
    }

    private UsageData? LoadCached()
    {
        try
        {
            if (!File.Exists(_cachePath)) return null;
            var json = File.ReadAllText(_cachePath);
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            var data = new UsageData();
            if (root.TryGetProperty("five_hour_percent", out var fh) && fh.ValueKind == JsonValueKind.Number)
                data.FiveHourPercent = fh.GetDouble();
            if (root.TryGetProperty("weekly_percent", out var wk) && wk.ValueKind == JsonValueKind.Number)
                data.WeeklyPercent = wk.GetDouble();
            if (root.TryGetProperty("five_hour_resets_at", out var fhr) && fhr.GetString() is string fhrStr)
                data.FiveHourResetsAt = DateTime.Parse(fhrStr, null, System.Globalization.DateTimeStyles.RoundtripKind);
            if (root.TryGetProperty("weekly_resets_at", out var wkr) && wkr.GetString() is string wkrStr)
                data.WeeklyResetsAt = DateTime.Parse(wkrStr, null, System.Globalization.DateTimeStyles.RoundtripKind);
            if (root.TryGetProperty("last_updated", out var lu) && lu.GetString() is string luStr)
                data.LastUpdated = DateTime.Parse(luStr, null, System.Globalization.DateTimeStyles.RoundtripKind);

            if (data.FiveHourPercent == null && data.WeeklyPercent == null) return null;
            return data;
        }
        catch { return null; }
    }
}
